# ã‚¹ãƒ†ãƒ¼ãƒˆãƒã‚·ãƒ³ãƒ»ãƒ“ãƒ˜ã‚¤ãƒ“ã‚¢ï¼ˆStatemachine Behaviourï¼›çŠ¶æ…‹é·ç§»æ©Ÿæ¢°ã®æŒ¯ã‚‹èˆã„ï¼‰
extends Node


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# ãƒãƒ¼ãƒ‰ãƒ»ãƒ‘ã‚¹é–¢é€£
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼


# çŒ¿å–å¾—
func monkey():
	return $"../ğŸµMonkey"


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# èµ·å‹•æ™‚è¨­å®š
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼


func _ready():
	self.monkey().statemachine().behaviour_of_entry = behaviour_of_entry
	self.monkey().statemachine().behaviour_of_exit = behaviour_of_exit


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# ã‚¤ãƒ™ãƒ³ãƒˆãƒãƒ³ãƒ‰ãƒ©
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼

# é–‹å§‹æ™‚ã®æŒ¯ã‚‹èˆã„
func behaviour_of_entry():
	# ãƒ­ãƒ¼ã‚«ãƒ«ã®ãƒ—ãƒ­ã‚°ãƒ©ãƒãƒ¼ãƒ»ãƒãƒ¼ãƒ‰ã¸ã‚¢ã‚¯ã‚»ã‚¹
	self.monkey().the_programmer_node().on_key_config_entered()

# çµ‚äº†æ™‚ã®æŒ¯ã‚‹èˆã„
func behaviour_of_exit():
	self.monkey().moderator().clear_count_by_step()
	
	# ç”»é¢è¡¨ç¤ºã€æ¼”å¥
	self.monkey().display().on_exit()

	# ã‚­ãƒ¼ã‚³ãƒ³ãƒ•ã‚£ã‚°å‘¼å‡ºã—å…ƒ
	self.monkey().owner_key_config_node().on_exit()
